File: algorithm/implementation/왕실의_나이트.java
package implementation;

import java.util.Scanner;

public class 왕실의_나이트 {
    /*
     * 나이트는 말을 타고 있기 때문에 이동을 할 때는 L자 형태로만 이동할 수 있으며 정원 밖으로는 나 갈 수 없다. 나이트는 특정한 위치에서 다음과 같은 2가지 경우로 이동할 수 있다．
     *
     * 1. 수평으로 두 칸 이동한뒤에 수직으로한칸 이동하기
     * 2. 수직으로 두 칸 이동한 뒤에 수평으로 한 칸 이동하기
     *
     * 이처럼 8 x8 좌표 평면상에서 나이트의 위치가 주어졌을 때 나이트가 이동할 수 있는 경우의 수를 출력하는 프로그램을 작성하시오.
     * 이때 왕실의 정원에서 행 위치를 표현할 때는 1부터 8로 표현하 며, 열 위치를 표현할 때는 a부터 h로 표현한다．
     *
     * 입력 조건)
     * 1) 첫째 줄에 8x8 좌표 평면상에서 현재 나이트가 위치한 곳의 좌표를 나타내는 두 문자로 구성된 문자 열이 입력된다. 입력 문자는 a1 처럼 열과 행으로 이뤄진다．
     * 2) 첫째 줄에 나이트가 이동할 수를 출력하시오，
     * */
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // 나이트의 위치 입력받기
        String input = scanner.nextLine();  //a1

        // 행, 열
        int row = input.charAt(1) - '0'; //1
        int col = input.charAt(0) - 'a' + 1; //a

        int result = 0;

        // x = row, y = col {x,y}
        int [][] steps = {
                {2,-1},{-2,-1},{2,1},{-2,-1} // 1. 수평으로 두  칸 이동한뒤에 수직으로한칸 이동하기
                ,{1,-2},{-1,-2},{1,2},{1,-2} // 2. 수직으로 두 칸 이동한 뒤에 수평으로 한 칸 이동하기
        };

        for (int i = 0; i < 8; i++) {
            // 이동하고자 하는 위치
            int next_row = row + steps[i][0];
            int next_col = col + steps[i][1];

            // 공간 벗어나면 무시
            if (next_col <= 8 && next_col >= 1 && next_row <= 8 && next_row >= 1) result ++;
        }

        System.out.println(result);
    }
}

File: algorithm/implementation/왕실의_나이트.kt
package implementation

import java.util.*

fun main(args: Array<String>) = with(Scanner(System.`in`)) {
    /*
    * 체스 판 : 8*8
    * 1. 수평으로 두 칸, 수직으로 한 칸
    * 2. 수직으로 두 칸, 수평으로 한 칸
    *
    * 나이트위 위치가 주어졌을 때 나이트가 이동할 수 있는 경우의 수 출력하시오
    * */

    // 나이트의 위치 입력 ex) a1
    val input = nextLine().toCharArray()
    var x = input[1].digitToInt()
    var y = input[0] - 'a' + 1

    // 이동할 수 있는 경우의 수 2차원 배열로 생성
    var steps = arrayOf(
          arrayOf(-1,-2)
        , arrayOf(-1,2)
        , arrayOf(1,-2)
        , arrayOf(1,2)
        , arrayOf(-2,-1)
        , arrayOf(-2,1)
        , arrayOf(2,-1)
        , arrayOf(2,1)
    )

    // 시뮬레이션
    var count = 0
    for (i in 0..7){
        var nx = x + steps[i][0]
        var ny = y + steps[i][1]

        if ((nx <= 8) and (nx >= 1) and (ny <= 8) and (ny >= 1)) count++
    }
    println(count)
}